packages/react-sdk/docusaurus/docs/React/06-ui-cookbook/13-custom-call-layout.mdx

```diff
@@ -342,17 +345,100 @@ const hasScreenShare = (p: StreamVideoParticipant) =>
+
+```tsx
+export const SpeakerView = () => {
+  const call = useCall();
+  const [participantInSpotlight, ...otherParticipants] = useParticipants();
+  // highlight-start
+  const [participantsBar, setParticipantsBar] = useState<HTMLDivElement | null>(
+    null,
+  );
+  // highlight-end
+
+  // determine whether the call is a 1:1 call
+  const isOneToOneCall = otherParticipants.length === 1;
+  useEffect(() => {
+    if (!call) return;
+    const customSortingPreset = getCustomSortingPreset(isOneToOneCall);
+    call.setSortParticipantsBy(customSortingPreset);
+  }, [call, isOneToOneCall]);
+
+  // highlight-start
+  useEffect(() => {
+    if (!participantsBar || !call) return;
+
+    const cleanup = call.viewportTracker.setViewport(participantsBar);
+
+    return () => cleanup();
+  }, [participantsBar, call]);
+  // highlight-end
+
+  return (
+    <div className="speaker-view">
+      {call && otherParticipants.length > 0 && (
+        // highlight-next-line
+        <div ref={setParticipantsBar} className="participants-bar">
+          {otherParticipants.map((participant) => (
+            <div className="participant-tile" key={participant.sessionId}>
+              <ParticipantView participant={participant} />
+            </div>
+          ))}
+        </div>
+      )}
 
-And that's it. We have successfully built our custom layout.
-At last, you can find the running example of this guide [here](https://github.com/GetStream/stream-video-js/sample-apps/react/cookbook-participant-list).
+      <div className="spotlight">
+        {call && participantInSpotlight && (
+          <ParticipantView
+            participant={participantInSpotlight}
+            videoKind={
+              hasScreenShare(participantInSpotlight) ? 'screen' : 'video'
+            }
+          />
+        )}
+      </div>
+
+      <CustomCallControls>
+        {call && <ScreenShareButton call={call} />}
+        <SpeakingWhileMutedNotification>
+          <ToggleAudioPublishingButton />
+        </SpeakingWhileMutedNotification>
+        <ToggleCameraPublishingButton />
+        {call && (
+          <CancelCallButton
+            call={call}
+            onLeave={() => {
+              console.log('onLeave callback called');
+            }}
+          />
+        )}
+      </CustomCallControls>
+    </div>
+  );
+};
+
+const CustomCallControls = ({ children }: PropsWithChildren<{}>) => {
+  return <div className="str-video__call-controls">{children}</div>;
+};
+
+const hasScreenShare = (p: StreamVideoParticipant) =>
+  p.publishedTracks.includes(SfuModels.TrackType.SCREEN_SHARE);
+```
```

Comment:
This is a continuation of Oliver's guide and I've decided to follow his style of "copy pasting" the whole code blocks and highlighting the newly added lines. This style makes it so that I have to write less descriptive text for each snippet e.g.: "take this effect and add it in the scope of the component, then create stateful element reference like this add it here..." - meaning you scroll and you see what has been added where and when you're at the end you can easily piece together what and why's instead of copying snippets and jumping from web to your code editor. It's harder to maintain and "uglier" but I feel like it "flows" nice if you know what I mean. @oliverlaz - what's your opinion?

Reply:
In this case, I think we should keep the whole code in a single block. I see the benefit in letting the user copy-paste stuff. Smaller code blocks, in this case, are increasing the potential efforts a reader might need to do in order to compile the code pieces together.

Reply:
Good catch, I'm not sure if call controls should even live within the layout component. I guess we could extract at least that. 👍

Reply:
Adjusted in: [fb11703](https://github.com/GetStream/stream-video-js/pull/511/commits/fb1170333b31bbd0f8823cf286ee5dfc95274e52)